components/playlist/browser/playlist_background_webcontents.h

```diff
@@ -0,0 +1,89 @@
+
+#ifndef BRAVE_COMPONENTS_PLAYLIST_BROWSER_PLAYLIST_BACKGROUND_WEBCONTENTS_H_
+#define BRAVE_COMPONENTS_PLAYLIST_BROWSER_PLAYLIST_BACKGROUND_WEBCONTENTS_H_
+
+#include <map>
+#include <memory>
+#include <vector>
+
+#include "base/containers/unique_ptr_adapters.h"
+#include "base/gtest_prod_util.h"
+#include "base/memory/weak_ptr.h"
+#include "base/time/time.h"
+#include "base/timer/timer.h"
+#include "brave/components/playlist/browser/playlist_media_handler.h"
+#include "brave/components/playlist/common/mojom/playlist.mojom.h"
+#include "url/gurl.h"
+
+FORWARD_DECLARE_TEST(CosmeticFilteringPlaylistFlagEnabledTest,
+                     AllowCosmeticFiltering);
+
+namespace content {
+class BrowserContext;
+class WebContents;
+}  // namespace content
+
+namespace playlist {
+
+class PlaylistService;
+
+// `PlaylistBackgroundWebContents` fulfills background `WebContents` requests.
+// After creating the background `WebContents`, it waits 10 seconds for the
+// first non-empty media list to arrive. On receiving the media, or if the timer
+// goes off (whichever happens first), it destructs the background
+// `WebContents`, and calls the provided callback with the result.
+// It overrides the user agent if `features::kPlaylistFakeUA` is enabled,
+// or uses a static look-up table to decide if it has to otherwise.
+class PlaylistBackgroundWebContents final {
+ public:
+  PlaylistBackgroundWebContents(content::BrowserContext* context,
+                                PlaylistService* service);
+  PlaylistBackgroundWebContents(const PlaylistBackgroundWebContents&) = delete;
+  PlaylistBackgroundWebContents& operator=(
+      const PlaylistBackgroundWebContents&) = delete;
+  ~PlaylistBackgroundWebContents();
+
+  void Add(const GURL& url,
+           PlaylistMediaHandler::OnceCallback on_media_detected_callback,
+           base::TimeDelta timeout = base::Seconds(10));
+
+  void Reset();
+
+ private:
+  FRIEND_TEST_ALL_PREFIXES(::CosmeticFilteringPlaylistFlagEnabledTest,
+                           AllowCosmeticFiltering);
+  FRIEND_TEST_ALL_PREFIXES(PlaylistBackgroundWebContentsTest,
+                           ExtractPlaylistItemsInTheBackground);
+  FRIEND_TEST_ALL_PREFIXES(PlaylistBackgroundWebContentsTest,
+                           UserAgentOverride);
+
+  void Remove(content::WebContents* web_contents,
+              PlaylistMediaHandler::OnceCallback on_media_detected_callback,
+              GURL url,
+              std::vector<mojom::PlaylistItemPtr> items);
+
+  // used by
+  // PlaylistBackgroundWebContentsTest.ExtractPlaylistItemsInTheBackground, and
+  // PlaylistBackgroundWebContentsTest.UserAgentOverride
+  content::WebContents& web_contents() const {
+    CHECK(background_web_contents_.size() == 1);
+    return *background_web_contents_.cbegin()->first;
+  }
+
+  raw_ptr<content::BrowserContext> context_;
+  raw_ptr<PlaylistService> service_;
+  std::map<std::unique_ptr<content::WebContents>,
+           base::OneShotTimer,
```

Comment:
The other callback belongs to `PlaylistMediaHandler` (and not `PlaylistBackgroundWebContentsHelper`), so we'd still have to split the `base::OnceCallback<>`, but I also think it's more expressive to have the timers in `PlaylistBackgroundWebContents`, as it keeps the logic of `WebContents` addition and removal together as much as possible.

Reply:
IMO having `PlaylistBackgroundWebContentsHelper` and `PlaylistMediaHandler ` as separate things does look like overkill. Does it actually simplify things and makes the logic clearer? Does it protect you from some misuses here and there?

Before you explained that, it was unclear to me that `PlaylistBackgroundWebContentsHelper` creates `PlaylistMediaHandler` implicitly. I think it might be a sign that there are just too much tab helpers in the Playlist machinery.

Reply:
`PlaylistBackgroundWebContentsHelper` exists so that we can separately define the concerns of a helper that's only attached to background `WebContents`es, and hence clearly see what we do differently (e.g. script injection) compared to when working with `WebContents`es that are browser tabs. On the other hand, `PlaylistMediaHandler` is a separate tab helper, as it's needed by both `PlaylistTabHelper` and `PlaylistBackgroundWebContentsHelper`, and therefore `PlaylistMediaHandler` is not mergeable with  `PlaylistBackgroundWebContentsHelper`.